python/node_similarity.py

```diff
@@ -0,0 +1,285 @@
+from math import sqrt
+from typing import Set, List, Callable, Tuple, Union
+from itertools import product
+import mgp
+
+# not implemented for directed graphs
+# not implemented for multigraphs
+# not taking into account edge weights
+
+neighbors_dict = dict()
+
+
+class Mode:
+    """
+    Constants for the mode parameter
+    """
+
+    PAIRWISE = ["p", "pairwise"]
+    CARTESIAN = ["c", "cartesian"]
+
+
+@mgp.read_proc
+def jaccard(
+    context: mgp.ProcCtx, node1: mgp.Any, node2: mgp.Any, mode: str = "cartesian"
+) -> mgp.Record(node1=mgp.Vertex, node2=mgp.Vertex, similarity=float):
+    """This procedure calls a method for calculating the similarity between two nodes (or two tuples of nodes)
+    so that the used node similarity measure is the Jaccard similarity.
+
+    :param node1: A node or a tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: A node or a tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param mode: Can be `p` (or `pairwise`) for pairwise similarity calculation or `c` (or `cartesian`)
+    for calculating the similarity of the Cartesian product of given tuples.
+    The default value is set to be a Cartesian product.
+    :type mode: str, optional
+
+    :return: This procedure returns a mgp.Record with 3 fields:
+    the first node, the second node and the Jaccard similarity between them
+    :rtype: mgp.Record
+    """
+
+    return _calculate_similarity(node1, node2, _calculate_jaccard, mode)
+
+
+@mgp.read_proc
+def overlap(
+    context: mgp.ProcCtx, node1: mgp.Any, node2: mgp.Any, mode: str = "cartesian"
+) -> mgp.Record(node1=mgp.Vertex, node2=mgp.Vertex, similarity=mgp.Number):
+    """This procedure calls a method for calculating the similarity between two nodes (or two tuples of nodes)
+    so that the used node similarity measure is the overlap similarity.
+
+    :param node1: A node or a tuple of nodes
+    :type node1: mgp.Vertex, tuple
+    :param node2: A node or a tuple of nodes
+    :type node2: mgp.Vertex, tuple
+    :param mode: Can be `p` (or `pairwise`) for pairwise similarity calculation or `c` (or `cartesian`)
+    for calculating the similarity of the Cartesian product of given tuples.
+    The default value is set to be a Cartesian product.
+    :type mode: str, optional
+
+    :return: This procedure returns a mgp.Record with 3 fields:
+    the first node, the second node and the overlap similarity between them
+    :rtype: mgp.Record
+    """
+
+    return _calculate_similarity(node1, node2, _calculate_overlap, mode)
```

Comment:
Here you can maybe even extract mode, so it is not `p` or `pairwise`, but that it is just `pairwise`. Same for `c` or `cartesian`. That way you have a clean state, and you can save it maybe as attribute, like current_mode

Reply:
I don't understand what the problem here is.